crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs

```diff
@@ -452,53 +457,46 @@ fn derive_libp2p_keypair(schnorrkel_sk: &schnorrkel::SecretKey) -> Keypair {
 /// previous segments to see if they are already in the cache. If they are not, they are added
 /// from DSN.
 async fn populate_pieces_cache<PG, PC>(
-    mut run_cache_sync_rx: Receiver<SegmentIndex>,
+    segment_index: SegmentIndex,
     piece_getter: Arc<FarmerPieceGetter<PG, PC>>,
     piece_cache: Arc<tokio::sync::Mutex<FarmerPieceCache>>,
 ) where
     PG: PieceGetter + Send + Sync,
     PC: PieceCache + Send + 'static,
 {
-    let piece_cache = piece_cache.clone();
-    let piece_getter = piece_getter.clone();
-
-    if let Some(segment_index) = run_cache_sync_rx.recv().await {
-        // Notify sender that we received a new segment index and started a cache sync.
-        run_cache_sync_rx.close();
-
-        info!(new_segment_index=%segment_index, "Started syncing piece cache...");
-
-        for piece_index in 0..segment_index.first_piece_index().into() {
-            let key = PieceIndex::from(piece_index).hash().to_multihash().into();
-
-            let should_cache = { piece_cache.lock().await.should_cache(&key) };
-            if should_cache {
-                let result = piece_getter
-                    .get_piece(piece_index.into(), PieceGetterRetryPolicy::Limited(1))
-                    .await;
-
-                match result {
-                    Ok(Some(piece)) => {
-                        debug!(%piece_index, "Added piece to cache.");
-                        piece_cache.lock().await.add_piece(key, piece);
-                    }
-                    Ok(None) => {
-                        debug!(%piece_index, "Couldn't find piece.");
-                    }
-                    Err(err) => {
-                        debug!(error=%err, %piece_index, "Failed to get piece for piece cache.");
-                    }
+    info!(%segment_index, "Started syncing piece cache...");
+    let final_piece_index =
+        u64::from(segment_index.first_piece_index()) + ArchivedHistorySegment::NUM_PIECES as u64;
+
+    for piece_index in 0..final_piece_index {
```

Comment:
Hm... this means we'll start from scratch on every restart and it will take some time to go through all indices. Let's have at least a TODO since I expect this to be a usability problem.

Also I believe you meant this:
```suggestion
    for piece_index in 0..=final_piece_index {
```

Reply:
The naming is the issue indeed. It is not the final index, it is the index right after last index. I'd use `(0..).take(num_piece_indices)` instead to clearly show what you want to do here.

Reply:
The code changed. I keep `final_piece_index for now.